arangod/RocksDBEngine/RocksDBTempStorage.cpp

```diff
@@ -0,0 +1,265 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2022 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Julia Puget
+////////////////////////////////////////////////////////////////////////////////
+
+#include "RocksDBTempStorage.h"
+
+#include "Basics/FileUtils.h"
+#include "Basics/VelocyPackHelper.h"
+#include "Basics/application-exit.h"
+#include "Basics/debugging.h"
+#include "Basics/files.h"
+#include "Logger/LogMacros.h"
+#include "RestServer/TemporaryStorageFeature.h"
+#include "RocksDBEngine/RocksDBColumnFamilyManager.h"
+#include "RocksDBEngine/RocksDBSortedRowsStorageContext.h"
+
+#ifdef USE_ENTERPRISE
+#include "Enterprise/RocksDBEngine/EncryptionProvider.h"
+#include "Enterprise/RocksDBEngine/RocksDBEncryptionUtilsEE.h"
+#endif
+
+#include <absl/strings/str_cat.h>
+
+#include <rocksdb/comparator.h>
+#include <rocksdb/db.h>
+#include <rocksdb/env.h>
+#include <rocksdb/env_encryption.h>
+#include <rocksdb/options.h>
+#include <rocksdb/slice_transform.h>
+
+using namespace arangodb;
+
+namespace {
+
+class KeysComparator : public rocksdb::Comparator {
+ public:
+  KeysComparator() = default;
+  int Compare(rocksdb::Slice const& lhs,
+              rocksdb::Slice const& rhs) const override {
+    TRI_ASSERT(lhs.size() >= sizeof(uint64_t));
+    TRI_ASSERT(rhs.size() >= sizeof(uint64_t));
+    // compare first 8 bytes, which is the "context id"
+    int diff = memcmp(lhs.data(), rhs.data(), sizeof(uint64_t));
+    if (diff != 0) {
+      return diff;
+    }
+
+    // move beyond "context id"
+    auto p1 = lhs.data() + sizeof(uint64_t);
+    auto p2 = rhs.data() + sizeof(uint64_t);
+
+    // now compare the running number (used for stable sorts)
+    int diffInId = 0;
+    bool hasPrefixId1 = static_cast<size_t>(p1 - lhs.data()) < lhs.size();
+    bool hasPrefixId2 = static_cast<size_t>(p2 - rhs.data()) < rhs.size();
+    if (hasPrefixId1 && hasPrefixId2) {
+      diffInId = memcmp(p1, p2, sizeof(uint64_t));
+      p1 += sizeof(uint64_t);
+      p2 += sizeof(uint64_t);
+    } else if (hasPrefixId1) {
+      diffInId = 1;
+    } else if (hasPrefixId2) {
+      diffInId = -1;
```

Comment:
Should we not advance p2 by sizeof(uint64_t) in this case?

Reply:
When we arrive here, the lhs and rhs differ in size, so we wouldn't reach the while below to catch the actual key value and the order in which to compare
```
 while (static_cast<size_t>(p1 - lhs.data()) < lhs.size() &&
           static_cast<size_t>(p2 - rhs.data()) < rhs.size())
 ```
 meaning the string with lower byte size will always score lower, so we don't have to advance the pointer for the rest of the data and just return the result